knowhere/index/vector_index/IndexDiskANN.cpp

```diff
@@ -0,0 +1,269 @@
+
+        if (prep_conf.use_bfs_cache) {
+            pq_flash_index_->cache_bfs_levels(prep_conf.num_nodes_to_cache, node_list);
+        } else {
+            pq_flash_index_->generate_cache_list_from_sample_queries(
+                warmup_query_file, 15, 6, prep_conf.num_nodes_to_cache, prep_conf.num_threads, node_list);
+        }
+        pq_flash_index_->load_cache_list(node_list);
+    }
+
+    // set thread number
+    omp_set_num_threads(prep_conf.num_threads);
+
+    // warmup
+    if (prep_conf.warm_up) {
+        LOG_KNOWHERE_INFO_ << "Warming up.";
+        uint64_t warmup_L = 20;
+        uint64_t warmup_num = 0;
+        uint64_t warmup_dim = 0;
+        uint64_t warmup_aligned_dim = 0;
+        T* warmup = nullptr;
+        if (file_exists(warmup_query_file)) {
+            diskann::load_aligned_bin<T>(warmup_query_file, warmup, warmup_num, warmup_dim, warmup_aligned_dim);
+        } else {
+            warmup_num = (std::min)((_u32)150000, (_u32)15000 * prep_conf.num_threads);
+            warmup_dim = pq_flash_index_->get_data_dim();
+            warmup_aligned_dim = ROUND_UP(warmup_dim, 8);
+            diskann::alloc_aligned(((void**)&warmup), warmup_num * warmup_aligned_dim * sizeof(T), 8 * sizeof(T));
+            std::memset(warmup, 0, warmup_num * warmup_aligned_dim * sizeof(T));
+            std::random_device rd;
+            std::mt19937 gen(rd());
+            std::uniform_int_distribution<> dis(-128, 127);
+            for (uint32_t i = 0; i < warmup_num; ++i) {
+                for (uint32_t d = 0; d < warmup_dim; ++d) {
+                    warmup[i * warmup_aligned_dim + d] = (T)dis(gen);
+                }
+            }
+        }
+        std::vector<uint64_t> warmup_result_ids_64(warmup_num, 0);
+        std::vector<float> warmup_result_dists(warmup_num, 0);
+
+#pragma omp parallel for schedule(dynamic, 1)
+        for (_s64 i = 0; i < (int64_t)warmup_num; ++i) {
+            pq_flash_index_->cached_beam_search(warmup + (i * warmup_aligned_dim), 1, warmup_L,
+                                                warmup_result_ids_64.data() + (i * 1),
+                                                warmup_result_dists.data() + (i * 1), 4);
+        }
+        if (warmup != nullptr) {
+            diskann::aligned_free(warmup);
+        }
+    }
+
+    is_prepared_ = true;
+    return true;
+}
+
+template <typename T>
+DatasetPtr
+IndexDiskANN<T>::Query(const DatasetPtr& dataset_ptr, const Config& config,
+                       const faiss::BitsetView /* not implemented yet */) {
+    CheckPreparation(is_prepared_);
+
+    auto query_conf = DiskANNQueryConfig::Get(config);
+    auto& k = query_conf.k;
+
+    GET_TENSOR_DATA_DIM(dataset_ptr);
+    auto query = static_cast<T*>(p_data);
+    auto p_id_u64 = std::vector<uint64_t>(k * rows).data();
+    auto p_id = static_cast<int64_t*>(malloc(sizeof(int64_t) * k * rows));
+    auto p_dist = static_cast<float*>(malloc(sizeof(float) * k * rows));
```

Comment:
Would be better to use new and delete[] or simply vector to avoid deallocating resources manually?

Reply:
> Would be better to use new and delete[] or simply vector to avoid deallocating resources manually?

Must use malloc here, because free is used when Dataset destruct